### ISSUE TITLE
Add a uses_request decorator for XML RPC methods

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3073]
|<img alt="amcnabb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| amcnabb reported|
|-|-|
|Trac ID|trac#3073|
|Type|enhancement|
|Created|2008-03-11 01:29:22Z|
|Branch|https://github.com/twisted/twisted/tree/xmlrpc-request-object-3073|

It is often useful for XML RPC methods to have access to the request object.  For example, they may need to find the client's host or look into HTTP headers.  The provided patch makes this simple.

If an XML RPC method has an attribute called uses_request, then the request will be passed as the first argument (after self).  The uses_request decorator adds this attribute to a method.

Attachments:

* [uses_request.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4b5/4b5553b2666ee42e1a757900e4ca0b4e79d28703/9b541d706306d14b7629e9af569b5a773879df62.patch) (1609 bytes) - added by amcnabb on 2008-03-11 20:34:13Z - 
* [withRequest.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4b5/4b5553b2666ee42e1a757900e4ca0b4e79d28703/5b792e4a2bd86380e48ad4e67e0f3e95396f8c79.patch) (3517 bytes) - added by wam on 2010-11-05 10:22:34Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3073 3073
type__enhancement enhancement
reporter__amcnabb amcnabb
priority__normal normal
milestone__ 
branch__branches_xmlrpc_request_object_3073 branches/xmlrpc-request-object-3073
branch_author__exarkun__Daniel_Kraft exarkun, Daniel Kraft
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1205198962000000 1205198962000000
changetime__1289231938000000 1289231938000000
version__None None
owner__ 
cc__wam
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml set owner to amcnabb|
|-|-|

I agree, this is a useful feature to have -- thanks for submitting the patch. However, I think you should take a different approach with the decorator implementation.

In the patch you provide, the "pass in the request" logic is split between the decorator and the base class. This is not ideal, as it creates a new, separate and hidden interface between the XML-RPC resource and the decorator.

But I'm scratching around for alternatives. The best two I can come up with are:
 1. to have an alternative prefix for methods that take a request object.
 1. to have a different base class that always passes request in.

The first one is kind of similar to your patch, in that `XMLPRC` will still have an `if` check. The second is cleaner, but will make it a little less nice to mix methods that need `request` and those that don't.

I think I prefer the second option. It will require some refactoring to make an easy point for subclassing and overriding. One way to do this is to override the `_getFunction` class, decorating its result with something that passes the request object in. Of course, that will mean that `_getFunction` needs to grow a `request` parameter.

Let me know what you think.

At another level, this patch needs tests and a couple of coding standard changes (we use `camelCase` and put a newline after the first `"""` in a docstring).

Thanks again for submitting this,
jml

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml commented|
|-|-|

Also, I forgot to mention.

A lot of newcomers to testing can find it hard to write automated tests for their patches. If you need any help with writing the tests, please ask here, on the mailing list or on #twisted. We'll be happy to lend a hand.

**Comment by twisted-trac** [comment]
|<img alt="Dwig's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|Dwig commented|
|-|-|

I suggest looking at the problem from a bit larger perspective.  In my situation, I want to look at the request on the way in, before dispatching to the function or the subhandler, in order to authenticate it.  I've seen example code that overrides render() to do authentication, but I'd like to see an interface more directly oriented toward handling the request object.  Maybe something like a chain of responsibility, where the main request handler and subhandlers can register something resembling "event callbacks".

**Comment by twisted-trac** [comment]
|<img alt="amcnabb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|amcnabb commented|
|-|-|

Thanks for looking at the patch.

I don't like either of the two alternatives.  The first alternative, having an alternative prefix, seems confusing.  Two methods could be the same except for the prefix, and then you would have to have clearly defined (but easily forgotten) precedence rules.  Plus it clutters things more.  Python's xmlrpclib doesn't have a prefix at all, and Twisted's approach was a breath of fresh air.  Everything is a normal method unless it starts with "xmlrpc_".  Period.

I don't like the second alternative because it makes it impossible to decide whether to input request on a method-by-method level.  I am using this in a real world application, and it seems that most methods don't require the request.  It's very nice to specify it on a method-by-method basis.  Also, if a brand new method is added which requires the request, it seems silly to have to change the base class and rewrite all of the other methods.

I understand your point about there being a new interface between the XML-RPC resource and the decorator, but I take a different perspective.  I think that the interface is between the XML-RPC resource and its method.  The decorator is just a shortcut for giving an attribute to the method.

**Comment by twisted-trac** [comment]
|<img alt="amcnabb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|amcnabb commented|
|-|-|

This updated patch switches to camelCase and adds documentation to XMLRPC's docstring.

**Comment by twisted-trac** [comment]
|<img alt="feisley's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|feisley commented|
|-|-|

I think a combination of these approaches might be beneficial. I am involved in the PyAMF project and we had a similar issue come up where we wanted to pass the request object from twisted to the handler methods. We accomplished this by adding an attribute to the class (expose_request). Additionally you can decorate a handler class with the @expose_request decorator. In this way the main processor looks first to the method and then to the class to see if it should inject the request.

As an aside, this can also be slightly adapted to support some form of class wide and/or method preprocessor or authenticator to act before the handler method is called.

One way or another we would like to see some form of request injection added as currently we have overridden the entire render_POST() method to implement these elements (expose_request and authenticator) to the XMLRPC module.

**Comment by twisted-trac** [comment]
|<img alt="wam's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|wam commented|
|-|-|

I just implemented the same thing in the same way after I had a discussion about that on #twisted. A decorator per method was the proposed solution.
It just happened, that I implemented this before I found this bug.

The implementation of this is pretty easy and simple to test (see patch). The resource class is also easy to understand.
I'd be willing to expand the functionality so that a object-variable would also trigger the request exposure to all methods in it. I'd set this object variable with a class decorator in the same way as the method decorator in the appended patch.

I also wouldn't want another base-class. Another prefix seems less explicit to me. @withRequest has a docstring where another prefix doesn't.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [30256]) Branching to 'xmlrpc-request-object-3073'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [30257]) Apply withRequest.patch

refs [#3073](https://github.com/twisted/twisted/issues/3073)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @exarkun|
|-|-|

It seems to me that fixing this at all would be better than continuing to wait for a perfect solution to come along (particularly as none has for the last three years).  I think the decorator-based approach is fine, nor does it preclude someone from doing something better later on.  So...

  1. `functools.wraps` isn't available in Python 2.4, so we can't use it.  Fortunately your code didn't end up using it, so we can just delete the import. :)
  1. The indentation on the example needs to be a little different for epytext to accept it: example text should be indented a bit, and preceded with a `::`.
  1. Top-level suites (eg function definitions) should be separated by three blank lines (so more whitespace after `withRequest`, for example)
  1. The two new methods in the test suite should have docstrings
  1. Whenever you have something async happening in a test method, it's likely that you need to return a Deferred from that test method.  Otherwise the Deferred might fail and trial might not notice.
  1. There isn't any real test coverage that extra arguments are still passed to `withRequest` decorated methods.  Passing one extra argument in the test would be a good way to exercise that.
  1. Also needs a [news file](http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles) describing the change

There's nothing too controversial here, so I'll make the changes and land the fix.

Thanks for pushing this forward wam.  And thanks for the original patch amcnabb, sorry it took so long.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [30258]) Address review feedback

refs [#3073](https://github.com/twisted/twisted/issues/3073)

**Comment by twisted-trac** [comment]
|<img alt="wam's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|wam commented|
|-|-|

Thanks for the feedback and the additional work!
I'll fix [#4729](https://github.com/twisted/twisted/issues/4729) according to your guidelines and upload the patch again for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [30259]) Merge xmlrpc-request-object-3073

Author: wam
Reviewer: exarkun
Fixes: [#3073](https://github.com/twisted/twisted/issues/3073)

Add a `withRequest` decorator to `twisted.web.xmlrpc` to allow
exposed methods access to the HTTP request object.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
